analyzers/tests/SonarAnalyzer.Test/Helpers/DisjointSetsPrimitivesTest.cs

```diff
@@ -0,0 +1,70 @@
+﻿using static SonarAnalyzer.Helpers.DisjointSetsPrimitives;
+
+namespace SonarAnalyzer.Test.Helpers;
+
+[TestClass]
+public class DisjointSetsPrimitivesTest
+{
+    [TestMethod]
+    public void FindRootAndUnion_AreConsistent()
```

Comment:
Because the way in which parents are updated is an implementation detail of the data structure. The test asserts the coherency of how those two work with each other, which is all that is needed for the data structure to work properly.
Looking into the dictionary is breaking that abstraction level. 

I could hide the parent dictionary into a dedicated interface that would not give access to the underlying data, but that would increase the complexity for no reason.

Having the test looking in the internals of how disjoint sets updates parents would also be problematic in extensions. There are two extensions that we are likely going to make to this data structure, for performance reasons: [union by rank](https://en.wikipedia.org/wiki/Disjoint-set_data_structure#Union_by_rank) and [path compression](https://en.wikipedia.org/wiki/Disjoint-set_data_structure#Union_by_rank:~:text=There%20are%20several%20algorithms%20for%20Find%20that%20achieve%20the%20asymptotically%20optimal%20time%20complexity.%20One%20family%20of%20algorithms%2C%20known%20as%20path%20compression%2C%20makes%20every%20node%20between%20the%20query%20node%20and%20the%20root%20point%20to%20the%20root.%20Path%20compression%20can%20be%20implemented%20using%20a%20simple%20recursion%20as%20follows%3A).

If I look into the parents and assert the specific behavior we need to rewrite all tests at every improvement of the data structure.

Reply:
Makes sense.
Thanks Anton for the explanation.